source/NVDAObjects/UIA/__init__.py

```diff
@@ -1172,12 +1172,24 @@ def _get_states(self):
 			states.add(controlTypes.STATE_INVALID_ENTRY)
 		if self._getUIACacheablePropertyValue(UIAHandler.UIA_IsRequiredForFormPropertyId):
 			states.add(controlTypes.STATE_REQUIRED)
+
+		# Read-only state
 		try:
-			isReadOnly=self._getUIACacheablePropertyValue(UIAHandler.UIA_ValueIsReadOnlyPropertyId,True)
+			isReadOnly = self._getUIACacheablePropertyValue(UIAHandler.UIA_ValueIsReadOnlyPropertyId, True)
 		except COMError:
-			isReadOnly=UIAHandler.handler.reservedNotSupportedValue
-		if isReadOnly and isReadOnly!=UIAHandler.handler.reservedNotSupportedValue:
+			isReadOnly = UIAHandler.handler.reservedNotSupportedValue
+		if (
+			isReadOnly == UIAHandler.handler.reservedNotSupportedValue
+			and self.UIATextPattern
+		):
+			# Most UIA text controls don't support the "ValueIsReadOnly" property, so we need to look at
+			# 	root document "IsReadOnly" attribute.
+			document = self.UIATextPattern.documentRange
+			isReadOnly = document.GetAttributeValue(UIAHandler.UIA_IsReadOnlyAttributeId)
+		# We can check "isReadOnly" again
+		if isReadOnly:

```

Comment:
I think reservedNotSupportedValue evaluates to True. I think we either need to do

```suggestion
		if isReadOnly and isReadOnly != UIAHandler.handler.reservedNotSupportedValue:
```

or

```suggestion
		if isReadOnly is True:
```